src/components/SelectionController.tsx

```diff
@@ -0,0 +1,292 @@
+      newSelection = { ...oldSelection };
+      if (oldSelected === true) delete newSelection[file.id];
+      else if (file.selectable !== false) newSelection[file.id] = true;
+      break;
+    case SelectionType.Range:
+      let indexA = prevIndex as number;
+      let indexB = displayIndex;
+      if (indexA > indexB) [indexA, indexB] = [indexB, indexA];
+      for (let i = indexA; i < indexB + 1; ++i) {
+        const file = files[i];
+        if (!isNil(file) && file.selectable !== false)
+          newSelection[file.id] = true;
+      }
+      break;
+  }
+
+  return {
+    selection: newSelection,
+    previousSelectionIndex: selectionIndexToPersist,
+  };
+};
+
+const reducer: React.Reducer<State, Action> = (state, action) => {
+  switch (action.type) {
+    case 'new_files': {
+      const {
+        selection: oldSelection,
+        previousSelectionIndex: prevIndex,
+      } = state;
+      if (isNil(oldSelection)) {
+        return { selection: {}, previousSelectionIndex: undefined };
+      }
+
+      const { files } = action;
+
+      const selection = {};
+      let previousSelectionIndex = undefined;
+      if (isArray(files)) {
+        previousSelectionIndex = isNumber(prevIndex)
+          ? clampIndex(prevIndex, files)
+          : undefined;
+        files.map(file => {
+          if (!isObject(file)) return;
+          const wasSelected = oldSelection[file.id] === true;
+          const canBeSelected = file.selectable !== false;
+          if (wasSelected && canBeSelected) selection[file.id] = true;
+        });
+      }
+
+      return { selection, previousSelectionIndex };
+    }
+    case 'new_folder_chain':
+      return { selection: {}, previousSelectionIndex: undefined };
+    case 'disable_selection':
+      return { selection: {}, previousSelectionIndex: undefined };
+    case 'set_selection': {
+      const { selection = {}, selectionIndex } = action;
+      return { selection, previousSelectionIndex: selectionIndex };
+    }
+    case 'file_single_click': {
+      const { event } = action;
+
+      try {
+        let type = isMobileDevice()
+          ? SelectionType.Multiple
+          : SelectionType.Single;
+        if (event.ctrlKey || event.key === KbKey.Space)
+          type = SelectionType.Multiple;
+        if (event.shiftKey) type = SelectionType.Range;
+
+        return handleSelectionToggle(action, state, type);
+      } catch (error) {
+        ConsoleUtil.logInternalException(error, 'handling selection toggle.');
+      }
+    }
+    case 'new_sort': {
+      const { files = [] } = action;
+      const { selection } = state;
+
+      const newSelection = {};
+      for (const file of files) {
+        if (isNil(file) || selection[file.id] !== true) continue;
+        newSelection[file.id] = true;
+      }
+
+      return { selection: newSelection, previousSelectionIndex: undefined };
+    }
+    default:
+      return state;
+  }
+};
+
+export const SelectionController = ({
```

Comment:
I think this can be typed as `SelectionController: FC<Props> = ({ . . . }) => { . . . }`. Also, can you rename `Props` into `StateControllerProps` to make it easier to reuse the type in tests and such?

Reply:
Sounds fine with me.  I chose to type it the way I did to make it clear that the `children` prop would not be as typed by `React.FC`, which says children is a `ReactNode`.  I can do what you say though, since the `children` type will be overwritten by my `StateControllerProps` interface.